File: libs/api-client/src/problem.ts
/** The header that carries the flow identifier of a request. See `ERR-006`. */
export const FLOW_ID_HEADER = 'X-Flow-ID';

/** The media type of every error response of the hub. */
export const PROBLEM_MEDIA_TYPE = 'application/problem+json';

/** One field that caused a rejection. */
export interface FieldFailure {
  /** The reason, as a sentence a person reads. */
  detail: string;
  /** A JSON Pointer in the fragment form of RFC 6901, so it starts with `#`. */
  pointer: string;
}

/** The body of an error response. It obeys RFC 9457. */
export interface Problem {
  /** The problem type, a relative reference of the form /problems/<owner>/<slug>. */
  type: string;
  title: string;
  status: number;
  detail?: string;
  instance?: string;
  errors?: FieldFailure[];
}

/** The problem types that the hub owns. */
export const PROBLEM_TYPE = {
  requestInvalid: '/problems/http/request-invalid',
  bodyInvalid: '/problems/http/body-invalid',
  accessDenied: '/problems/http/access-denied',
  networkNotAllowed: '/problems/hub/network-not-allowed',
  notFound: '/problems/http/not-found',
  settingsAbsent: '/problems/hub/settings-absent',
  methodNotAllowed: '/problems/http/method-not-allowed',
  identityLast: '/problems/hub/identity-last',
  validationFailed: '/problems/http/validation-failed',
  settingsInvalid: '/problems/hub/settings-invalid',
  internal: '/problems/http/internal'
} as const;

/**
 * Reports whether the value carries the three members that a problem requires.
 * A body that another server wrote reaches this too, so it checks the shape.
 */
export function isProblem(value: unknown): value is Problem {
  if (typeof value !== 'object' || value === null) {
    return false;
  }

  const candidate = value as Partial<Problem>;

  return (
    typeof candidate.type === 'string' &&
    typeof candidate.title === 'string' &&
    typeof candidate.status === 'number'
  );
}

File: libs/rest/problem.go
package rest

// ProblemMediaType is the content type of every error response. The name
// carries the subject, because this module also answers application/json.
const ProblemMediaType = "application/problem+json"

// Problem is the body of an error response.
type Problem struct {
	Type     string         `json:"type"`
	Title    string         `json:"title"`
	Status   int            `json:"status"`
	Detail   string         `json:"detail,omitempty"`
	Instance string         `json:"instance,omitempty"`
	Errors   []FieldFailure `json:"errors,omitempty"`
}

// FieldFailure names one field that caused a rejection.
type FieldFailure struct {
	Detail  string `json:"detail"`
	Pointer string `json:"pointer"`
}

// WithDetail returns the problem with the detail. A detail names this occurrence
// and never carries the text of an error, a query, or a secret.
func (p Problem) WithDetail(detail string) Problem {
	p.Detail = detail

	return p
}

// WithErrors returns the problem with one item for each field that failed.
func (p Problem) WithErrors(failures ...FieldFailure) Problem {
	p.Errors = failures

	return p
}

// WithStatus returns the problem with another status. A caller that holds a
// status and no cause uses it, such as the one that rewrites a failure of a
// transport it does not own.
func (p Problem) WithStatus(status int) Problem {
	p.Status = status

	return p
}

// newProblem builds a problem of one registered type.
func newProblem(problemType, title string, status int) Problem {
	return Problem{Type: problemType, Title: title, Status: status}
}
